### PR TITLE
Add Adafruit esp32s2 boards with custom bootloader, partition and upload.extra_flags

### PR DESCRIPTION
**Pull request by hathach**
done on behalf of @ladyada, this PR does:
- Add [Adafruit FunHouse](https://www.adafruit.com/product/4985) and upcoming [Adafruit Feather Esp32s2](https://blog.adafruit.com/2021/01/01/panel-mount-tft-esp32-s2-feather)
- Update Adafruit magtag and metro esp32s2 to have tinyuf2 as factory app
- Also modify platform.txt as discussed in  adafruit#2  +  adafruit#3 +  adafruit#4 to add support for 
  - Board/variant partition.csv, bootloader.bin ( priority is user sketch > variant > build.parition/boot )
  - upload.extra_flags to optionally flash tinyuf2 as factory application
  - Remove duplicated parameter in running python script in linux and windows

Please review and let me know if you want any changes. 

Notes: [tinyuf2](https://github.com/adafruit/tinyuf2) is a uf2 bootloader for multiple platforms. For esp32s2, it is an factory app to perform uf2 firmware update.

### CODE REVIEW

File: platform.txt [hunk 1]
```diff
@@ -94,11 +94,18 @@ build.loop_core=
 build.event_core=
 build.extra_flags=-DESP32 -DCORE_DEBUG_LEVEL={build.code_debug} {build.loop_core} {build.event_core} {build.defines} {build.extra_flags.{build.mcu}}
 
-# Check if custom partitions exist
-recipe.hooks.prebuild.1.pattern=bash -c "[ ! -f "{build.source.path}"/partitions.csv ] || cp -f "{build.source.path}"/partitions.csv "{build.path}"/partitions.csv"
-recipe.hooks.prebuild.2.pattern=bash -c "[ -f "{build.path}"/partitions.csv ] || cp "{runtime.platform.path}"/tools/partitions/{build.partitions}.csv "{build.path}"/partitions.csv"
+# Check if custom partitions exist: source > variant > build.partitions
+recipe.hooks.prebuild.1.pattern=bash -c "[ ! -f {build.source.path}/partitions.csv ] || cp -f {build.source.path}/partitions.csv {build.path}/partitions.csv"
```
Comment (me-no-dev) [inline platform.txt:98]: please make sure that you escape all path variables. It is causing an issue when paths have spaces :)

File: platform.txt [hunk 2]
```diff
@@ -94,11 +94,18 @@ build.loop_core=
...
-recipe.hooks.prebuild.2.pattern.windows=cmd /c if not exist "{build.path}\partitions.csv" COPY "{runtime.platform.path}\tools\partitions\{build.partitions}.csv" "{build.path}\partitions.csv"
+recipe.hooks.prebuild.2.pattern.windows=cmd /c if not exist "{build.path}\partitions.csv" if exist "{build.variant.path}\partitions.csv" COPY "{build.variant.path}\partitions.csv" "{build.path}\partitions.csv"
+recipe.hooks.prebuild.3.pattern.windows=cmd /c if not exist "{build.path}\partitions.csv" COPY "{runtime.platform.path}\tools\partitions\{build.partitions}.csv" "{build.path}\partitions.csv"
+
+# Check if custom bootloader exist: source > variant > build.boot
+recipe.hooks.prebuild.4.pattern=bash -c "[ -f {build.source.path}/bootloader.bin ] && cp -f {build.source.path}/bootloader.bin {build.path}/{build.project_name}.bootloader.bin || ( [ -f {build.variant.path}/bootloader.bin ] && cp {build.variant.path}/bootloader.bin {build.path}/{build.project_name}.bootloader.bin || cp -f {runtime.platform.path}/tools/sdk/{build.mcu}/bin/bootloader_{build.boot}_{build.flash_freq}.bin {build.path}/{build.project_name}.bootloader.bin )"
```
Comment (me-no-dev) [inline platform.txt:107]: Path variables also need escaping here

File: platform.txt (hunk 1 above)
Comment (hathach) [reply platform.txt]: sorry, my bad, fixed in latest commit

File: platform.txt (hunk 2 above)
Comment (hathach) [reply platform.txt]: thanks, fixed in latest commit

**Comment by hathach** [comment]
merge latest commit from master to resolve platform.txt conflict, set base branch to other and back to get github update file changes.

**Comment by me-no-dev** [comment]
Thanks @hathach :) merged

**Comment by UnexpectedMaker** [comment]
This is cool!  @hathach does this mean I can duplicate this with the TinyS2 and FeatherS2 variant + UF2 firmware and it will also work for buy boards?

**Comment by hathach** [comment]
> Thanks @hathach :) merged

Superb! Thank you for a quick review and merge.

> This is cool! @hathach does this mean I can duplicate this with the TinyS2 and FeatherS2 variant + UF2 firmware and it will also work for buy boards?

yeah, you will need partition.csv bootloader.bin and tinyuf2.bin . You could find that in the tinyuf2 release or build it yourself. Just follow my changes for adafruit boards.

**Comment by me-no-dev** [comment]
@UnexpectedMaker I will suggest you hurry up :) I need to release 2.0.0-rc1 to the board manager and will be nice if it's in there

**Comment by UnexpectedMaker** [comment]
> @UnexpectedMaker I will suggest you hurry up :) I need to release 2.0.0-rc1 to the board manager and will be nice if it's in there

Ok, please don't wait for me then. I need a few days to get something implemented and tested and to truly understand the implications of doing this... like that I assume I'd lose the ability to allow people to pick custom partitions from the tools menu in the IDE etc. Some folks actually use that ;)

Maybe I'll have something in for a future RC release.

It's very exciting though!

**Pull request merged**